sound/soc/sof/core.c

```diff
@@ -255,22 +254,9 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	snd_sof_free_debug(sdev);
 dbg_err:
 	snd_sof_remove(sdev);
-#else
 
-	/*
-	 * when the probe_continue is handled in a work queue, the
-	 * probe does not fail so we don't release resources here.
-	 * They will be released with an explicit call to
-	 * snd_sof_device_remove() when the PCI/ACPI device is removed
-	 */
-
-fw_trace_err:
-fw_run_err:
-fw_load_err:
-ipc_err:
-dbg_err:
-
-#endif
+	sdev->fw_state = SOF_FW_BOOT_NOT_STARTED;
```

Comment:
@kv2019i I think we should set this in fw_unload instead of doing it here.

Reply:
I also feel like at this point the state should be set to PREPARE and not NOT_STARTED

Reply:
@ranj063 Hmm, I choose NOT_STARTED as the end-state is that we have removed (or the probe has failed) the DSP. I.e. the state is NOT_STARTED. 

Load error is just one failure. We need to also fix the case where DSP probe fails in prone_continue, and system goes to suspend. In this case. So to address that, if sof_probe_continue fails, we go back to NOT_STARTED and all resources are freed. 

I did think about putting the state variables to snd_sof_remove(), but there are only two calls to this function and putting this state logic in ops.h seemed less obvious than just "open coding" it to core.c. But yeah, maybe I'll put it to sof_remove still...

Reply:
@kv2019i I have an alternate suggestion. Instead of changing the fw state at this point, why not use the correct state in remove? look below for example

Reply:
@ranj063 Good point. I updated the series with an additional comment.